packages/apollo/src/commands/client/push.ts

```diff
@@ -1,70 +1,195 @@
-          this.debug(operationManifest);
+              const signatureToOperation = Object.fromEntries(
+                Object.entries(
+                  this.project.mergedOperationsAndFragmentsForService
+                ).map(([operationName, document]) => {
+                  const operationDefinition = document.definitions.find(
+                    ({ kind }) => kind === "OperationDefinition"
+                  );
+                  const relativePath =
+                    operationDefinition &&
+                    operationDefinition.loc &&
+                    relative(
+                      config.configURI ? config.configURI.fsPath : "",
+                      URI.parse(operationDefinition.loc.source.name).fsPath
+                    );
+                  const line =
+                    operationDefinition &&
+                    operationDefinition.loc &&
+                    operationDefinition.loc.source.locationOffset.line;
+                  return [
+                    operationHash(
+                      defaultOperationRegistrySignature(document, operationName)
+                    ),
+                    {
+                      operationName,
+                      document,
+                      file: line
+                        ? `${relativePath}:${line}`
+                        : relativePath || ""
+                    }
+                  ];
+                })
+              );
 
-          await project.engine.registerOperations(variables);
+              const { name, referenceID, version } = config.client!;
+              if (!name) {
+                throw new Error("Client name is required to push");
+              }
 
-          // store data for logging
-          ctx.operations = operationManifest;
-          ctx.serviceName = variables.id;
-          ctx.clientIdentity = variables.clientIdentity;
-        }
-      }
-    ]);
+              const variables = {
+                clientIdentity: {
+                  name: name,
+                  identifier: referenceID || name,
+                  version
+                },
+                id: config.name,
+                operations: operationManifest,
+                manifestVersion: 2
+              };
+              const { operations: _op, ...restVariables } = variables;
+              this.debug("Variables sent to Engine:");
+              this.debug(restVariables);
+              this.debug("Operations sent to Engine:");
+              this.debug(operationManifest);
+
+              const {
+                invalidOperations,
+                newOperations,
+                registrationSuccess
+              } = await project.engine.registerOperations(variables);
 
-    this.log(
-      `Successfully pushed ${operations.length} operations from the ${
-        clientIdentity.name
-      } client to the ${serviceName} service in Engine`
-    );
+              if (!registrationSuccess) {
+                if (invalidOperations) {
```

Comment:
Is it possible for `!registrationSuccess && !invalidOperations`? I only ask in case it's a possible scenario, and if it has any semantics that we can communicate.

I don't imagine so, but I'd suggest that we throw or log something if that's the case.

Reply:
Right now it shouldn't be possible, though maybe in the future. I'll add an error message!